internal/verifier/oplog_reader.go

```diff
@@ -0,0 +1,716 @@
+
+	common.logger = v.logger
+	common.metaDB = v.metaClient.Database(v.metaDBName)
+
+	common.resumeTokenTSExtractor = oplog.GetRawResumeTokenTimestamp
+
+	o := &OplogReader{ChangeReaderCommon: common}
+
+	o.createIteratorCb = o.createCursor
+	o.iterateCb = o.iterateCursor
+
+	return o
+}
+
+func (o *OplogReader) createCursor(
+	ctx context.Context,
+	sess *mongo.Session,
+) (bson.Timestamp, error) {
+	savedResumeToken, err := o.loadResumeToken(ctx)
+	if err != nil {
+		return bson.Timestamp{}, errors.Wrap(err, "loading persisted resume token")
+	}
+
+	var allowDDLBeforeTS bson.Timestamp
+
+	var startTS bson.Timestamp
+
+	if token, has := savedResumeToken.Get(); has {
+		var rt oplog.ResumeToken
+		if err := bson.Unmarshal(token, &rt); err != nil {
+			return bson.Timestamp{}, errors.Wrap(err, "parsing persisted resume token")
+		}
+
+		ddlAllowanceResult := o.getMetadataCollection().FindOne(
+			ctx,
+			bson.D{
+				{"_id", o.ddlAllowanceDocID()},
+			},
+		)
+
+		allowanceRaw, err := ddlAllowanceResult.Raw()
+		if err != nil {
+			return bson.Timestamp{}, errors.Wrap(err, "fetching DDL allowance timestamp")
+		}
+
+		allowDDLBeforeTS, err = mbson.Lookup[bson.Timestamp](allowanceRaw, "ts")
+		if err != nil {
+			return bson.Timestamp{}, errors.Wrap(err, "parsing DDL allowance timestamp doc")
+		}
+
+		startTS = rt.TS
+	} else {
+		startOpTime, latestOpTime, err := oplog.GetTailingStartTimes(ctx, o.watcherClient)
+		if err != nil {
+			return bson.Timestamp{}, errors.Wrapf(err, "getting start optime from %s", o.readerType)
+		}
+
+		allowDDLBeforeTS = latestOpTime.TS
+
+		_, err = o.getMetadataCollection().ReplaceOne(
+			ctx,
+			bson.D{
+				{"_id", o.ddlAllowanceDocID()},
+			},
+			bson.D{
+				{"ts", allowDDLBeforeTS},
+			},
+			options.Replace().SetUpsert(true),
+		)
+		if err != nil {
+			return bson.Timestamp{}, errors.Wrapf(err, "persisting DDL-allowance timestamp")
+		}
+
+		startTS = startOpTime.TS
+
+		err = o.persistResumeToken(ctx, oplog.ResumeToken{startTS}.MarshalToBSON())
+		if err != nil {
+			return bson.Timestamp{}, errors.Wrap(err, "persisting resume token")
+		}
+	}
+
+	o.logger.Info().
+		Any("startReadTs", startTS).
+		Any("currentOplogTs", allowDDLBeforeTS).
+		Msg("Tailing oplog.")
+
+	sctx := mongo.NewSessionContext(ctx, sess)
+
+	findOpts := options.Find().
+		SetCursorType(options.TailableAwait)
+
+	if util.ClusterHasBSONSize([2]int(o.clusterInfo.VersionArray)) {
+		findOpts.SetProjection(o.getExprProjection())
+	}
+
+	oplogFilter := bson.D{{"$and", []any{
+		bson.D{{"ts", bson.D{{"$gte", startTS}}}},
+
+		bson.D{{"$expr", agg.Or{
+			// plain ops: one write per op
+			append(
+				agg.And{agg.In("$op", mslices.Of("d", "i", "u"))},
+				o.getNSFilter("$$ROOT")...,
+			),
+
+			// op=n is for no-ops, so we stay up-to-date.
+			agg.Eq{"$op", "n"},
+
+			// op=c is for applyOps, and also to detect forbidden DDL.
+			agg.And{
+				agg.Eq{"$op", "c"},
+				agg.Not{helpers.StringHasPrefix{
```

Comment:
[question] Do we limit ourselves to ignoring DDL operations on the "config" database because it would be more complex MQL to also ignore DDL operations on the "admin" database? I don't anticipate op=c oplog entries on the "config" database after a cluster is first initialized so I wanted to double check the intention here (I'm not proposing any changes).